purchase_update_proposal/models/purchase_order.py

```diff
@@ -0,0 +1,212 @@
+# -*- coding: utf-8 -*-
+# © 2021 David BEAL @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+from collections import defaultdict
+
+from openerp import _, api, fields, models
+from openerp.exceptions import Warning as UserError
+
+logger = logging.getLogger(__name__)
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    proposal_ids = fields.One2many(
+        comodel_name="purchase.line.proposal", inverse_name="order_id"
+    )
+    proposal_count = fields.Integer(compute="_compute_proposal")
+    proposal_state = fields.Selection(
+        selection=[
+            ("draft", "To Submit"),
+            ("submitted", "To Approve by an internal user"),
+            ("rejected", "Rejected"),
+            ("approved", "Approved"),
+        ],
+        readonly=True,
+        copy=False,
+        default="draft",
+    )
+    proposal_updatable = fields.Selection(
+        selection=[("", "undefined"), ("yes", "Yes"), ("no", "No")],
+        default="",
+        copy=False,
+        help="Computed to be sure that stock move are in a state "
+        "compatible with update",
+    )
+    proposal_date = fields.Date(
+        string="Update all dates at once",
+        help="Update all date of the order with this date",
+        copy=False,
+    )
+
+    def _check_updatable_proposal(self):
+        self = self.sudo()
+        for rec in self:
+            moves = self.env["stock.move"].search(
+                [("purchase_line_id", "in", rec.order_line.ids)]
+            )
```

Comment:
Easier to include `state in ('cancel', 'done')` in the search and skip the next check? (I mean, just check if `moves` is empty or not)